src/workerd/api/streams/standard.c++

```diff
@@ -2790,113 +2793,63 @@ ReadableStreamJsController::getController() {
-  using Readable = kj::Own<T>;
   using PartList = kj::Array<kj::ArrayPtr<byte>>;
 
-  AllReader(Readable readable, uint64_t limit)
-      : ioContext(tryGetIoContext()),
-        state(kj::mv(readable)),
-        limit(limit) {
-    state.template get<Readable>()->setOwner(addWeakRef());
-  }
-
-  void doClose(jsg::Lock& js) override {
-    if (isReadPending()) {
-      setPendingState(StreamStates::Closed());
-    } else {
-      state.template init<StreamStates::Closed>();
-    }
-  }
-
-  void doError(jsg::Lock& js, v8::Local<v8::Value> reason) override {
-    if (isReadPending()) {
-      setPendingState(js.v8Ref(reason));
-    } else {
-      state.template init<StreamStates::Errored>(js.v8Ref(reason));
-    }
-  }
+  AllReader(jsg::Ref<ReadableStream> stream, uint64_t limit)
+      : state(kj::mv(stream)),
+        limit(limit) {}
+  KJ_DISALLOW_COPY_AND_MOVE(AllReader);
 
   jsg::Promise<kj::Array<byte>> allBytes(jsg::Lock& js) {
-    auto after = [this](auto& js, PartList&& partPtrs) {
+    return loop(js).then(js, [this](auto& js, PartList&& partPtrs) {
       auto out = kj::heapArray<byte>(runningTotal);
       copyInto(out, kj::mv(partPtrs));
       return kj::mv(out);
-    };
-    auto promise = loop(js);
-    if (IoContext::hasCurrent()) {
-      return promise.then(js, IoContext::current().addFunctor(kj::mv(after)));
-    } else {
-      return promise.then(js, kj::mv(after));
-    }
+    });
   }
 
   jsg::Promise<kj::String> allText(jsg::Lock& js) {
-    auto after = [this](auto& js, PartList&& partPtrs) {
+    return loop(js).then(js, [this](auto& js, PartList&& partPtrs) {
       auto out = kj::heapArray<char>(runningTotal + 1);
       copyInto(out.slice(0, out.size() - 1).asBytes(), kj::mv(partPtrs));
       out.back() = '\0';
       return kj::String(kj::mv(out));
-    };
-    auto promise = loop(js);
-    if (IoContext::hasCurrent()) {
-      return promise.then(js, IoContext::current().addFunctor(kj::mv(after)));
-    } else {
-      return promise.then(js, kj::mv(after));
-    }
+    });
   }
 
-  size_t pendingReadCount = 0;
-  kj::Maybe<kj::OneOf<StreamStates::Closed, StreamStates::Errored>> maybePendingState = kj::none;
-
-  bool isReadPending() const {
-    return pendingReadCount > 0;
+  void visitForGc(jsg::GcVisitor& visitor) {
+    KJ_SWITCH_ONEOF(state) {
+      KJ_CASE_ONEOF(closed, StreamStates::Closed) {}
+      KJ_CASE_ONEOF(errored, StreamStates::Errored) {
+        visitor.visit(errored);
+      }
+      KJ_CASE_ONEOF(readable, jsg::Ref<ReadableStream>) {
+        visitor.visit(readable);
+      }
+    }
   }
 
 private:
-  kj::Maybe<IoContext&> ioContext;
-  kj::OneOf<StreamStates::Closed, StreamStates::Errored, Readable> state;
+  kj::OneOf<StreamStates::Closed, StreamStates::Errored, jsg::Ref<ReadableStream>> state;
   uint64_t limit;
   kj::Vector<jsg::BackingStore> parts;
   uint64_t runningTotal = 0;
+  kj::Maybe<kj::OneOf<StreamStates::Closed, StreamStates::Errored>> maybePendingState = kj::none;
```

Comment:
Given that it's only used here, I'd prefer to avoid using a typedef or using here. If I had to use this in a couple of places then I'd be fine with it.